src/sources/splunk_hec/acknowledgements.rs

```diff
@@ -0,0 +1,467 @@
+use futures_util::future::Shared;
+use roaring::RoaringTreemap;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    num::NonZeroU64,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc, Mutex,
+    },
+    time::{Duration, Instant},
+};
+use tokio::{sync::RwLock, time::interval};
+use vector_core::event::BatchStatusReceiver;
+use warp::Rejection;
+
+use crate::shutdown::ShutdownSignal;
+use crate::sources::util::finalizer::OrderedFinalizer;
+
+use super::ApiError;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(default)]
+pub struct HecAcknowledgementsConfig {
+    pub max_pending_acks: NonZeroU64,
+    pub max_number_of_ack_channels: NonZeroU64,
+    pub max_pending_acks_per_channel: NonZeroU64,
+    pub ack_idle_cleanup: bool,
+    pub max_idle_time: NonZeroU64,
+}
+
+impl Default for HecAcknowledgementsConfig {
+    fn default() -> Self {
+        Self {
+            max_pending_acks: NonZeroU64::new(10_000_000).unwrap(),
+            max_number_of_ack_channels: NonZeroU64::new(1_000_000).unwrap(),
+            max_pending_acks_per_channel: NonZeroU64::new(1_000_000).unwrap(),
+            ack_idle_cleanup: false,
+            max_idle_time: NonZeroU64::new(300).unwrap(),
+        }
+    }
+}
+
+pub struct IndexerAcknowledgement {
+    max_pending_acks: u64,
+    max_pending_acks_per_channel: u64,
+    max_number_of_ack_channels: u64,
+    channels: Arc<RwLock<HashMap<String, Arc<Channel>>>>,
+    shutdown: Shared<ShutdownSignal>,
+    total_pending_acks: AtomicU64,
+}
+
+impl IndexerAcknowledgement {
+    pub fn new(config: HecAcknowledgementsConfig, shutdown: Shared<ShutdownSignal>) -> Self {
+        let channels: Arc<RwLock<HashMap<String, Arc<Channel>>>> =
+            Arc::new(RwLock::new(HashMap::new()));
+        let max_idle_time = u64::from(config.max_idle_time);
+        let idle_task_channels = Arc::clone(&channels);
+
+        if config.ack_idle_cleanup {
+            tokio::spawn(async move {
+                let mut interval = interval(Duration::from_secs(max_idle_time));
+                loop {
+                    interval.tick().await;
+                    let mut channels = idle_task_channels.write().await;
+                    let expiring_channels = channels
+                        .iter()
+                        .filter_map(|(channel_id, channel)| {
+                            if channel.get_last_used().elapsed().as_secs() >= max_idle_time {
+                                Some(channel_id.clone())
+                            } else {
+                                None
+                            }
+                        })
+                        .collect::<Vec<String>>();
+
+                    for channel_id in expiring_channels {
+                        channels.remove(&channel_id);
+                    }
```

Comment:
I think this can all be replaced by `HashMap::retain`:
```suggestion
                    channels.retain(|channel_id, channel| channel.get_last_used().elapsed().as_secs() < max_idle_time);
```
I also note it would be worth caching `Instant::now()` before the loop and using `duration_since` instead of `elapsed`, as `elapsed` calls `gettimeofday` each time it's called.

Reply:
Oh cool, very clean. Will do.